File: 백준/Gold/1655. 가운데를 말해요/가운데를 말해요.cc
#include <bits/stdc++.h>

using namespace std;

int N;
priority_queue<int, vector<int>, less<>> Max;
priority_queue<int, vector<int>, greater<>> Min;

void Swap() {
	if (Max.empty()) return;
	if (Min.empty()) return;
	if (Max.top() <= Min.top()) return;
	int max = Max.top();
	int min = Min.top();
	Max.pop();
	Min.pop();
	Max.emplace(min);
	Min.emplace(max);
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);

	cin >> N;
	unsigned int num;

	for (int i = 0; i < N; i++) {
		cin >> num;
		if (Max.size() <= Min.size()) Max.emplace(num);
		else Min.emplace(num);
		Swap();
		printf("%d\n", Max.top());
	}

	return 0;
}

File: 백준/Gold/1655. 가운데를 말해요/가운데를 말해요.js
const fs = require('fs');
const file = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(file).toString().trim().split('\n');

const nums = input.slice(1).map(Number);

class PriorityQueue {
  constructor(compare) {
    this.heap = [null];
    this.compare = compare;
  }

  push(item) {
    this.heap.push(item);
    let current = this.heap.length - 1;
    let parent = Math.floor(current / 2);

    while (parent !== 0 && this.compare(this.heap[current], this.heap[parent])) {
      this._swap(parent, current);
      current = parent;
      parent = Math.floor(current / 2);
    }
  }

  pop() {
    if (this.heap.length === 1) return;
    if (this.heap.length === 2) return this.heap.pop();

    const top = this.heap[1];
    this.heap[1] = this.heap.pop();

    let [current, left, right] = [1, 2, 3];

    while (
      this.compare(this.heap[left], this.heap[current]) ||
      this.compare(this.heap[right], this.heap[current])
    ) {
      const topIndex = this.compare(this.heap[right], this.heap[left]) ? right : left;
      this._swap(current, topIndex);
      current = topIndex;
      left = current * 2;
      right = current * 2 + 1;
    }

    return top;
  }

  size() {
    return this.heap.length - 1;
  }

  top() {
    return this.heap[1];
  }

  _swap(a, b) {
    [this.heap[a], this.heap[b]] = [this.heap[b], this.heap[a]];
  }
}

const minHeap = new PriorityQueue((a, b) => a < b);
const maxHeap = new PriorityQueue((a, b) => a > b);
const result = [];

for (const num of nums) {
  if (maxHeap.size() <= minHeap.size()) maxHeap.push(num);
  else minHeap.push(num);

  if (minHeap.size() !== 0 && minHeap.top() < maxHeap.top()) {
    const max = maxHeap.pop();
    const min = minHeap.pop();
    maxHeap.push(min);
    minHeap.push(max);
  }

  result.push(maxHeap.top());
}

console.log(result.join('\n'));
